[Documentation] Fixtures customization guides - fixes (#10641)
Sylius/Sylius: docs/customization/fixtures.rst

    The former interface extends the ``ConfigurationInterface``, which is widely known from ``Configuration`` classes
    placed under ``DependencyInjection`` directory in Symfony bundles.

Why would you customize fixtures?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There are two main use cases for customizing fixture suites, in each of them you can adapt the data of your shop to be realistic,
the default fixtures suite of Sylius is selling clothes, if you are selling food you'd probably need your own fixtures to show that:

    * preparing test data for the development purposes like demo applications prepared for QA
    * preparing the shop configuration for the production instance

How to modify the existing Sylius fixtures?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> In Sylius, fixtures are added in ``src/Sylius/Bundle/CoreBundle/Resources/config/app/fixtures.yml``.

Reviewer: ```suggestion
In Sylius, fixtures are configured in ``src/Sylius/Bundle/CoreBundle/Resources/config/app/fixtures.yml`` file.
```

> We have the ``default`` suite that is partially-configured.

Reviewer: ```suggestion
It includes the ``default`` suite that is partially-configured.
```

> If you are planning to modify the default fixtures applied by the ``sylius:fixtures:load`` command, the only thing that you have

Reviewer: If you are planning to modify the default fixtures applied by the ``sylius:fixtures:load`` command, modify the ``config\packages\sylius_fixtures.yaml`` file.

                                        - "HUN_WEB"
                                    enabled: true

> Products are a more complicated fixture, because they have more dependencies (to Variants, Options etc.). In order to prepare a Product

Reviewer: ```suggestion
It is more complicated to create fixtures for products because they have more dependencies (to variants, options etc.). In order to prepare a product
```

you have to create not only the product itself but other related entities via their own factories.

> Sylius delivers four ready implementations of ``Product`` fixtures, that have their relevant options (like sizes for Tshirts):

Reviewer: ```suggestion
Sylius delivers four ready implementations of ``Product`` fixtures, that have their relevant options (like sizes for T-shirts):
```

* ``BookProductFixture``
* ``MugProductFixture``
* ``StickerProductFixture``
* ``TshirtProductFixture``

> You can modify their yaml fixture configs, but only within the capabilities delivered by thoses fixtures classes.

Reviewer: ```suggestion
You can modify their YAML fixture configs, but only within the capabilities delivered by those fixtures classes.
```

How to customize fixtures for customized models?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. tip::

>     The following example is based on other example of extending an entity with a new field.

Reviewer 2: Can we link to this example?

    You can browse the full implementation of this example on `this GitHub Pull Request
    <https://github.com/Sylius/Customizations/pull/23>`__.

> We the ``App\Entity\Shipping\ShippingMethod`` extended with a new field ``deliveryConditions``.

Reviewer: ```suggestion
Let's suppose we have extended ``App\Entity\Shipping\ShippingMethod``, added a new field ``deliveryConditions`` and want to include it in the fixtures.
```

> **1.** To cover that in fixtures, we will need  to override the ``ShippingMethodExampleFactory`` and add this field.

Reviewer: ```suggestion
**1.** To cover that in fixtures, we will need to override the ``ShippingMethodExampleFactory`` and add this field.
```

        private function getLocales(): iterable
        {
            /** @var LocaleInterface[] $locales */
            $locales = $this->localeRepository->findAll();
            foreach ($locales as $locale) {
                yield $locale->getCode();
            }
        }
    }

> **2.** Now we extend the ``Sylius\Bundle\CoreBundle\Fixture\ShippingMethodFixture`` in ``App\Entity\Fixture\ShippingMethodFixture``:

Reviewer: ```suggestion
**2.** Extend the ``Sylius\Bundle\CoreBundle\Fixture\ShippingMethodFixture`` in ``App\Entity\Fixture\ShippingMethodFixture``:
```

.. code-block:: php

    <?php

>     // src/Fixtures/ShippingMethodFixture.php

Reviewer: ```suggestion
    // src/Fixture/ShippingMethodFixture.php
```

        }
    }

> **3.** Then, in the ``config/services.yaml`` configure our services:

Reviewer: ```suggestion
**3.** Configure the services in ``config/services.yaml`` file:
```

.. code-block:: yaml

    sylius.fixture.example_factory.shipping_method:
        class: App\Fixture\Factory\ShippingMethodExampleFactory
        arguments:
            - "@sylius.factory.shipping_method"
            - "@sylius.repository.zone"
            - "@sylius.repository.shipping_category"
            - "@sylius.repository.locale"
            - "@sylius.repository.channel"
            - "@sylius.repository.tax_category"
        public: true

    sylius.fixture.shipping_method:
        class: App\Fixture\ShippingMethodFixture
        arguments:
            - "@sylius.manager.shipping_method"
            - "@sylius.fixture.example_factory.shipping_method"
        tags:
            - { name: sylius_fixtures.fixture }

.. tip::

>     When creating fixtures services manually, remember to turn off autowiring for them:

Reviewer: I don't think it's necessary (these fixtures can be fully autowired and autoconfigured with Sylius 1.5+).

Author: But we are in the 1.4 branch ;>

    .. code-block:: yaml

        App\:
            resource: '../src/*'
            exclude: '../src/{Entity,Fixture,Migrations,Tests,Kernel.php}'

> **4.** Finally, only what you have to do is add new Shipping Methods with delivery conditions in ``config/packages/fixtures.yaml``

Reviewer: ```suggestion
**4.** Add new Shipping Methods with delivery conditions in ``config/packages/sylius_fixtures.yaml``
```
